### ISSUE TITLE
[lit] Feature request: An easier way to use a WC in an env where other WCs may be registered

### ISSUE THREAD

**Issue by e111077** [lit/lit#2432]
## Description

Say you are writing a WC to be used in an environment where a lot of other WCs are defined. e.g. Home Assistant's plugin system.

If you want to use `@material/mwc-textfield` you have to check if `customElements.get('mwc-textfield')` exists and then _dynamically import_ the definition of `mwc-textfield`.

Scoped Element registries would help a lot here but are not a 1:1 solution here because:

If you want to extend the functionality of `mwc-textfield`, you can import `mwc-textfield-base` sans the custom element registration, and extend that definition and give it your own name. The problem now is that `mwc-textfield-base.js` also imports `@material/mwc-linear-ripple` and lots of other sub components, so you still get a custom element registry error thrown no matter what. The answer here is essentially go into the guts of everything, deep clone the files, and rewrite everything from scratch.

- Who wants the functionality
  - Users who may be working with bundles of WCs they may not own
  - Users trying to extend a web component
- What it is they want
  - An easier way to use an imported WC where you don't know if it's already registered
  - A way to extend a WC that uses another WC without having to rewrite the entire dep chain
- Why they want it
  - Sometimes there are environments where users do not control the bundles around them
- Functional description of task/subtask
  - Implement mitigation features
    - make it easier to publish components like adding an optional flag that checks the registry to
    `@customElement('my-el', {checkRegistry: true})`
    - publish a build plugin that can wrap custom element definition calls in registry checks
  - publish guidance on publishing elements and dealing with this issue
  - work with the WCCG for a community-based solution
  - help push a standards-based solution

## Acceptance criteria

Publish mitigation features, publish guidance on element authorship that tackles this, or prioritize support on a standards based level, or work with the WCCG for a community-based solution

**Comment by justinfagnani** [comment]
This is the real solution: https://github.com/WICG/webcomponents/issues/716
Public support for the feature on that issue would be helpful. We have a polyfill that we'll be putting some more work into this quarter.

**Comment by justinfagnani** [comment]
By the way, the problem you pointed out @e111077 of transitive dependencies needing to also use scoped registries is known - we're going to need the whole ecosystem to move over. Any other solution, like defensive registration or variable tag name would also require the entire set of dependencies to buy in.

The only other global option would be something like loading the library into an iframe, creating elements in that document, and moving them over - essentially using the iframe as an uber-scoped registry.

**Comment by e111077** [comment]
Yeah we had two separate issues come up in slack today regarding this. I know it's an ecosystem issue, but providing easy access to mitigation methods is one way to start pushing the ecosystem towards a workable direction

**Comment by anoblet** [comment]
The issue I had was Home Assistant comes bundled with their own version of Lit + a few Material Web Components.

I wanted to create a new card for Home Assistant which included `mwc-select`.

When importing `mwc-select`, Chrome complained that `mwc-ripple` had already been defined. Ignoring `@material/mwc-ripple` in rollup didn't work because it was imported as `@material/mwc-ripple/mwc-ripple` as opposed to `@material/mwc-list-item` which is imported using `@material/mwc-list`.

I then had to step through using two different plugins `rollup-plugin-exclude-dependencies-from-bundle` and `rollup-plugin-ignore` to first figure out _how_ the custom element was imported (adding it as a peer dependency, and excluding the whole namespace which resulted in import errors client-side), add them back, and then ignore/rewrite those specific imports.

If Home Assistant were to expose more of their frontend (node_modules) rather than bundling it in, this wouldn't be an issue. The user could upgrade/add new components without any conflict.

I was hoping to find a way where if I don't have ownership or control over the foundation, that there would be some external way to either scope new components and/or replace or upgrade definitions of already existing components.

This mostly seems to be an issue with design systems where two components share the same dependency, and upgrading/adding a component introduces a new version of that dependency. The developer then needs to upgrade all of the components within that design system (if it is possible) otherwise the new component will not work.

**Comment by justinfagnani** [comment]
This problem can occur in various host/plugin scenarios, so it'd be nice for us to figure out a solution even before scoped registries are ubiquitous.

> If Home Assistant were to expose more of their frontend (node_modules) rather than bundling it in, this wouldn't be an issue. The user could upgrade/add new components without any conflict.

Is there an issue for that? That would at least allow deduplication at the npm level if one were building Home Assistant and the card together. Would that even cover all cases though? I would assume that end users are installing pre-built packages and wouldn't get that out.

> I was hoping to find a way where if I don't have ownership or control over the foundation, that there would be some external way to either scope new components and/or replace or upgrade definitions of already existing components.

I think this is where using an iframe might work. We could try to see if the technique is workable and wrap it up in an easier to use utility library.

**Comment by guillemcordoba** [comment]
Just in case it's useful to anyone, this is the exact reason why I created [@scoped-elements/material-web](https://www.npmjs.com/package/@scoped-elements/material-web), which is just a re-export using the [@open-wc/scoped-elements](https://www.npmjs.com/package/@open-wc/scoped-elements) mixin.

Should also be compatible with @lit-labs/scoped-registry-mixin.

**Comment by robrez** [comment]
Needing scoped registrations has been one of the primary difficulties at my organization, where we have many teams vending components now.

As was pointed out in the issue above, one situation that is difficult is a component that depends on other components but does not itself scope those components.

For example, an app directly depends on `mwc-button@^v0.24.x`. That app depends on my component (we'll call it `fancy-widget`) which depends on `mwc-button@^0.25.x`.   Even if `fancy-widget` did scope `mwc-button`  its shadow root... `mwc-button` depends on `mwc-icon`, `mwc-ripple`.

So you have a problem where mwc-button itself must scope registrations (and so on down the line).. which is not common practice among design system vendors.

One of my naive (controversion) ideas is to register w/ a version. Probably at the "major" (left-most-non-zero) precision.  This might be in addition to or alternatively versus the standard agnostic registration - in either case as an (maybe anonymous) subclass.

```
<mwc-button-0-25>
  #
    <mwc-icon-0-25>
    <mwc-ripple-0-25>
```

Bring out the tar-and-feathers!  I am aware that the scoped registry solves the same problem.  The potential appeal to me has been:

- that this works w/o a scopedRegistry; and,
- solves an moderate concern area for me:  Portals;  eg: tooltips, popovers, dialogs

Many portal implementations append a portal onto document.body and provide a render fn or template as the means to hydrate the telported content.  It might be possible to pass the scoped registry to a lit-html renderer.. but, I'm just not confident that there wouldn't be a problem contending with the global registry.  I saw some mention of throwing exceptions when moving an element from one DocumentOrShadowRoot to another. In that same thread, someone made a super awesome [suggestion](https://github.com/WICG/webcomponents/issues/716#issuecomment-935619848)  .. but many of us need to take action sooner than later.

Talking through it, probably be better off just rendering into an "on-demand" shadowRoot everything teleported in this fashion... and teleporting that.   Creating and registering a custom-element anytime you want to teleport content wouldn't be ideal in some cases but would be fine for others.

**Comment by robrez** [comment]
@guillemcordoba

> Just in case it's useful to anyone, this is the exact reason why I created [@scoped-elements/material-web](https://www.npmjs.com/package/@scoped-elements/material-web), which is just a re-export using the [@open-wc/scoped-elements](https://www.npmjs.com/package/@open-wc/scoped-elements) mixin.
> 
> Should also be compatible with @lit-labs/scoped-registry-mixin.

Fantastic example, and thanks for sharing.  This is the kind of pattern we need when scoping 3rd party systems is necessary

**Comment by anoblet** [comment]
I opened a discussion topic in home-assistant/frontend: https://github.com/home-assistant/frontend/discussions/11423

**Comment by frank-dspeed** [comment]
@justinfagnani i guess the real solution is that the package maintainer the one of the real composit project is aware of all components and the used tags so he can on release or bundle creation dedub components. 

composing already pre registered components without name collision is only possible with hijacked tag names with hashes. 

what is your conclusion i think webcomponents should simply get shipped unregistered and should get registered if needed 
you can already only register a single class only once with the registrie as that class gets then completed with a tagName so that you can create new Element() instances programatical when you now register the same class with a other tagName what behavior do you expect? when i do new Element()

**Comment by ChristianUlbrich** [comment]
To whom it may concern. You can monkey-patch the CustomElements registry. I have created a [verrrry rough prototype](https://github.com/zalari/custom-elements-loader-prototype), allowing to register as many classes as you want for the **same** tag. The techniques applied in the prototype allow for a lot of different usage scenarios. Maybe someone finds the time to hack further on it and use it for some exciting stuff - basically running multiple versions of the same component using the same tag should be possible with this hack, because it simply wraps the globally unique tag and is reflecting content into specific versions of this tag.

**Comment by anoblet** [comment]
Home Assistant seems to be making the move to a scoped registry: https://github.com/custom-cards/boilerplate-card/pull/50

**Comment by justinfagnani** [comment]
The scoped element registries proposal is progressing. I don't think there's much good we can do here in the meantime, considering that we need to get transitive dependencies to adopt whatever strategy is used here.
